### ISSUE TITLE
Assertion error when running `rotatedRectangleIntersection`

### ISSUE THREAD

**Issue by jbwintergerst** [opencv/opencv#19824]
##### System information (version)

- OpenCV => 4.2.0.34 / 4.5.1.48 (python)
- Operating System / Platform => Ubuntu 20.04 (linux 5.8.0-48-generic), x86_64
- Compiler => N/A

##### Detailed description

OpenCV is throwing an assertion error when running `rotatedRectangleIntersection`:

```
cv2.error: OpenCV(4.2.0) /io/opencv/modules/imgproc/src/intersection.cpp:271: error: (-215:Assertion failed) fabs(normL2Sqr<float>(intersection[minI] - intersection[minJ]) - minD) < 1e-6 in function 'rotatedRectangleIntersection'
```

Issue has been discussed on the OpenCV forum [here](https://forum.opencv.org/t/assertion-error-when-running-rotatedrectangleintersection/2514).

> **TL;DR**
> Assertion error is solved by applying an offset to bring the rectangles closer to the (0, 0), which implies that maybe it is linked to the rectangle being "far away".

##### Steps to reproduce

```python
import cv2 as cv
import numpy as np


def box(box):
    return ((box[0], box[1]), (box[2], box[3]), box[4])


box_1 = np.array([ 246805.033  , 4002326.94   ,      26.40587,       6.20026,     -62.10156])
box_2 = np.array([ 246805.122  , 4002326.59   ,      27.4821 ,       8.5361 ,     -56.33761])

try:
    print(cv.rotatedRectangleIntersection(box(box_1), box(box_2)))  # Assertion error
except Exception as e:
    print(e)

offset = (246805, 4002326, 0, 0, 0)

box_1 -= offset
box_2 -= offset

print(cv.rotatedRectangleIntersection(box(box_1), box(box_2)))  # that works
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by gasparitiago** [comment]
@jbwintergerst:

Please test it and let me know if it solved the problem: https://github.com/opencv/opencv/pull/19842

**Comment by jbwintergerst** [comment]
Hi @gasparitiago ! Ty for the quick reaction and fix.
To test it, I presume their is no python build of this pull request.. Is their a way to use the python wrapper with a custom openCV C library version ?

**Comment by gasparitiago** [comment]
@jbwintergerst , to be very honest with you, I don't know if this is the best way for doing it, but here in my machine, I've built the opencv using cmake and adding some python flags: https://stackoverflow.com/questions/37070304/how-to-build-opencv-for-python3-when-both-python2-and-python3-are-installed/39409570

After executing the commands:

```
make
make install
```

the cv2 package of my python already pointed to the local version of opencv that I've built.

**Comment by alalek** [comment]
> make install

Be careful with that (especially with "sudo"). There is no robust "uninstall" step.

To run python script with configured OpenCV binaries from "build" directory you can use this statement instead:

```
./setup_vars.sh python3 <your script>
```

**Comment by jbwintergerst** [comment]
I'm testing inside a docker container, so I'm not that concerned about uninstalling things ^^
thx for the advice, I'll test that asap !

**Comment by jbwintergerst** [comment]
@gasparitiago I can confirm that your fix works for me as well in my use case

**Comment by alalek** [comment]
To avoid such numeric calculation errors we can apply input data normalization in prolog and epilog of `rotatedRectangleIntersection` implementation.

This is proposed in the comments on the forum:

> It might be a solution to offset the rectangles closer to the (0, 0) before calculating the iou.

```
Point2f shift_pt = (rect1.center + rect2.center) * 0.5;
```

---

Question about the improvement of raised assertion check.
Proposals are welcome

---

Also we need C++ regression test for the case from this issue

**Comment by gasparitiago** [comment]
@alalek 

I will try to implement the solution proposed in the comments when I got some time, but I have one question:

> Also we need C++ regression test for the case from this issue

Is there a document explaining how to perform this kind of tests so I can take a look at it?

**Comment by alalek** [comment]
We have an example of similar regression test: https://github.com/opencv/opencv/blame/4.5.2/modules/imgproc/test/test_intersection.cpp#L335-L350

**Comment by gasparitiago** [comment]
@alalek 

Updated the https://github.com/opencv/opencv/pull/19842 with the normalization and included a regression test.
